only use accurate powf function

vtjnash opened the pull request:
The powi intrinsic optimization over calling powf is that it is inaccurate.
We don't need that.

When it is equally accurate (e.g. tiny constant powers),
LLVM will already recognize and optimize any call to a function named `powf`,
and produce the same speedup.

fix #19872

vtjnash commented:
Note that this reverts https://github.com/JuliaLang/julia/issues/2741#issuecomment-47008166 (e.g. restores https://github.com/JuliaLang/julia/commit/f2b3192eb2e59dd34e1e2838036a13167410814f), since being correct is more important than being a few % faster.

JeffBezanson commented on test/math.jl:
```diff
@@ -959,9 +959,13 @@ end
 @testset "issue #13748" begin
-    let A = [1 2; 3 4]; B = [5 6; 7 8]; C = [9 10; 11 12]
-        @test muladd(A,B,C) == A*B + C
-    end
```
Did you mean to delete this?

yuyichao commented on src/codegen.cpp:
```diff
@@ -5964,7 +5961,7 @@ static void init_julia_llvm_env(Module *m)
         &pow,
```
This breaks compilation here.

What's the reason again that we can't do the `static_cast` unconditionally? Including `<math.h>` in C++ code causes `using std::pow;` which provides additional overloads for `float` and `long double`.

yuyichao replied:
More precisely what `glibc` + `libstdc++`/`libc++` do when `<math.h>` is directly included by c++ code is roughly

```c++
extern double pow(double, double);

namespace std {
using ::pow;
float pow(float, float);
long double pow(long double, long double);
// template version for promotion...
}

using std::pow;
```

It shouldn't be possible to overwrite a previously defined `::pow(double, double)` so I expect MSVC to do something similar and it should be safe in general to specify the overload type using a `static_cast`.

vtjnash replied:
My main concern would be whether those are the same function. I would expect that the C++ version is getting name mangled might not end up picking up the openlibm version.

yuyichao replied:
I don't think that is an issue.

1. These should be inline functions that calls the C `::pow`, `::powf`, `::powl`.
2. For C++ implementations that provide these it shouldn't be allowed to overwrite the existing `::pow(double, double)` symbol so the it should always be the C one.
3. At least in the generic linux binary `cglobal(:pow)` and `ccall(:pow, ..)` points to the `libm` version.

StefanKarpinski commented:
Are we moving ahead with this? If not, we should remove the 0.6.0 label. If so, it would be good to get the tests passing...

vtjnash commented:
That's the plan, just made an error in rebasing. But it also now made a good excuse to try out the new `llvmcall` support.

vchuravy commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

tkelman commented:
replutil test failure on win32 is real - probably related to #17251, will see if locally running with `--precompiled=no` makes it pass

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19890#issuecomment-274725188) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/a00a4a716b498ec9cd3b42e8f4925687674aebe7/74776c0_vs_06fa32c/report.md). cc @jrevels

StefanKarpinski commented:
Another failure in `test/replutil.jl`...?

JeffBezanson commented:
`test/replutil` tends to generate action-at-a-distance test failures, since it looks at e.g. what is printed for method tables of Base functions. Change some methods and the output changes.

tkelman commented:
This failure is very specific to debug info in backtraces for the pow function. Not action at a distance at all, for this PR.

vtjnash commented:
>  [1] nan_dom_err; at .\math.jl:311 [inlined]

looks like the name-demanger is getting confused here, nothing to do with #17251

vtjnash commented:
> possible performance regressions were detected

Just a note that these performance reductions are both expected and intentional. It's good that we have a record of them though, and will probably want to link back here / note this on the nightly as well.

KristofferC commented:
Is it expected that `exp` became a lot slower as well? I thought we had a pure julia version of that right now.

tkelman commented:
We should probably revert exp back to using openlibm then, if changing the pow calculation used by julia is going to be this expensive. Or change its code to opt in to the other pow version, if it isn't needed for accuracy in the innards of the exp implementstion? cc @musm

KristofferC commented:
The julia code for `exp` seemed very well tested for accuracy so it should be possible to keep the performance for `exp` which is where a lot of the slowdowns are being reported.

tkelman commented:
could `@fastmath` be usable for allowing users to annotate whether they prefer accuracy or speed for this?

vtjnash commented:
yes, `FastMath.pow` implements the fast (previous) version

vtjnash commented:
It looks like `exp` was just computing `2 ^ constant`, and my change was stopping it from doing constant propagation / inlining. Easy fix.

tkelman commented:
great @nanosoldier `runbenchmarks(ALL, vs=":master")`

musm commented on base/math.jl:
```diff
@@ -677,14 +679,11 @@ function modf(x::Float64)
-^(x::Float32, y::Int32) = powi_llvm(x, y)
-^(x::Float16, y::Integer) = Float16(Float32(x)^y)
+@inline ^(x::Float64, y::Float64) = nan_dom_err(ccall("llvm.pow.f64", llvmcall, Float64, (Float64, Float64), x, y), x + y)
+@inline ^(x::Float32, y::Float32) = nan_dom_err(ccall("llvm.pow.f32", llvmcall, Float32, (Float32, Float32), x, y), x + y)
```
I might have missed something, but why now llvm for `^` when both arguments are floats? the libm pow function is accurate.

vtjnash replied:
It resolves to the same function, but give an optimization hint for llvm that we mean *that* pow, and not just some random function named pow.

musm commented:
https://github.com/JuliaLang/julia/blob/e577f8e763d81fd8fa24e37859881d45c145b857/base/math.jl#L508
Also needs  `^ = Base.FastMath.pow_fast `  for similar reason. Should trigger a regression.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19890#issuecomment-275744024) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/c55b9b4ecdc84f73fb916d5696d0181d4d539514/441405c_vs_8091c21/report.md). cc @jrevels

tkelman commented:
ldexp uses pow in the subnormal case, which maybe nanosoldier didn't hit? looks like it's only powers of two so should also use the fast version

vtjnash commented:
I also fixed pow to do constant propagation, so it doesn't show up anymore.

tkelman commented:
Where and how? That sounds easy to break if it isn't tested and tracked.

vtjnash commented:
I added `@inline` so the inliner would stop getting tripped up by the cold branch for error checking. We have perf testing for this now.

tkelman commented:
including the subnormal branch of ldexp?

@nanosoldier `runbenchmarks(ALL, vs=":master")`

vtjnash commented:
What's that got to do with anything? Nanosoldier was already run before I rebased this today (https://github.com/JuliaLang/julia/pull/19890#issuecomment-275796933)

tkelman commented:
@musm requested that you use the fast pow in ldexp which you have not done

nanosoldier commented:
Something went wrong when running [your job](https://github.com/JuliaLang/julia/pull/19890#issuecomment-275980514):
```
NanosoldierError: failed to run benchmarks against primary commit: ErrorException("failed process: Process(`sudo cset shield -e su nanosoldier -- -c ./benchscript.sh`, ProcessExited(1)) [1]")
```
Logs and partial data can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/e6157010d42bbc53a36031945702b0751be2cb70/a45b4f9_vs_65afe47/report.md)
cc @jrevels

jrevels commented:
Nanosoldier is failing on master right now, so the failure here is probably unrelated to this PR (ref [my comment here](https://github.com/JuliaLang/julia/commit/65afe479cfd5afd6c782aafc2a18dac5a55cb4de#commitcomment-20657791))

KristofferC commented:
Are we waiting for nanosoldier to work again to check before this is merged?

tkelman commented:
waiting for ldexp and exp to use a consistent choice of pow between the two. if fast pow is no longer necessary in exp either then yes nanosoldier will be needed to verify that

tkelman commented:
I think the issue with Compat shouldn't be a problem when nanosoldier runs against latest master
@nanosoldier `runbenchmarks(ALL, vs=":master")`

vtjnash commented on base/special/exp.jl:
```diff
@@ -114,13 +118,13 @@ function exp{T<:Union{Float32,Float64}}(x::T)
             # add significand_bits(T) + 1 to lift the range outside the subnormals
             twopk = reinterpret(T, rem(exponent_bias(T) + significand_bits(T) + 1 + k, fpinttype(T)) << significand_bits(T))
-            return y*twopk*T(2.0)^(-significand_bits(T) - 1)
+            return y * twopk * T(2.0)^(-significand_bits(T) - 1)
```
should this also be `eps(T) / 2`?

musm replied:
I think it's better as is, since the intent of the code is a lot more clear with it being explicit same with `ldexp`  (which is why I didn't use `eps`, as it makes the scaling and explicit) 

but if they must change it would also make sense to change both 
`T(2.0)^(exponent_max(T) - 1)`
and
`T(2.0)^(-significand_bits(T) - 1)`
should either both us the `^ = Base.FastMath.pow_fast` or be evaluated ahead as constants. 

I have to admit it's a little disappointing that we have to now work around these  cases as such

vtjnash replied:
> I have to admit it's a little disappointing that we have to now work around these cases as such

All of these options generate exactly the same result. Just trying to understand what should be in the code for the most clarity.

simonbyrne replied:
@musm I'm still not sure I quite understand the issue here. All these expressions are still being constant folded. e.g. on this branch:
```
julia> import Base.significand_bits

julia> foo(T) = T(2.0)^(-significand_bits(T) - 1)
foo (generic function with 1 method)

julia> @code_llvm foo(Float64)

define double @julia_foo_65393(i8**) #0 !dbg !5 {
L12:
  ret double 0x3CA0000000000000
}

julia> @code_llvm foo(Float32)

define float @julia_foo_65426(i8**) #0 !dbg !5 {
L12:
  ret float 0x3E70000000000000
}
```

simonbyrne replied:
@musm I'm still not sure I quite understand the issue here. All these expressions are still being constant folded. e.g. on this branch:
```
julia> import Base.significand_bits

julia> foo(T) = T(2.0)^(-significand_bits(T) - 1)
foo (generic function with 1 method)

julia> @code_llvm foo(Float64)

define double @julia_foo_65393(i8**) #0 !dbg !5 {
L12:
  ret double 0x3CA0000000000000
}

julia> @code_llvm foo(Float32)

define float @julia_foo_65426(i8**) #0 !dbg !5 {
L12:
  ret float 0x3E70000000000000
}
```

simonbyrne replied:
(so this means 1282205a0285ffacd431f5e0a13cbcf8524468f2 is unnecessary, and can be removed).

tkelman replied:
that commit seems like an improvement in readability to me

musm replied:
> @musm I'm still not sure I quite understand the issue here. All these expressions are still being constant folded. e.g. on this branch:

I'm a little confused then why we need:
https://github.com/JuliaLang/julia/pull/19890/files#diff-4da275bf96d6e97c1918f46754df04d1R68
in the `exp` function since 
https://github.com/JuliaLang/julia/blob/1282205a0285ffacd431f5e0a13cbcf8524468f2/base/special/exp.jl#L121
and 
https://github.com/JuliaLang/julia/blob/1282205a0285ffacd431f5e0a13cbcf8524468f2/base/special/exp.jl#L127
should also be constant folded (i.e. `T(2.0)^(exponent_max(T) - 1)` and  `T(2.0)^(-significand_bits(T) - 1)` ). So doesn't that make the`    ^ = Base.FastMath.pow_fast` unnecessary  (the benchmarks showed otherwise...?)  (here's the def for `exponent_max` https://github.com/JuliaLang/julia/blob/1282205a0285ffacd431f5e0a13cbcf8524468f2/base/float.jl#L751)

@tkelman 
> that commit seems like an improvement in readability to me
The reason I suggested the current code is more clear is that it's really obvious what's currently happening with the `signficand_bits` and the scaling, e.g.  see the line and the following one
https://github.com/JuliaLang/julia/blob/master/base/math.jl#L556
it may be less readable, but I believe it's more clear and would be  a lot easier to quickly infer what it's doing upon first glance or in the future when revisiting the logic. 

This type of scaling is also used in almost the exact same logic, repeatedly, in the `exp` function https://github.com/JuliaLang/julia/blob/1282205a0285ffacd431f5e0a13cbcf8524468f2/base/special/exp.jl#L118
I'd prefer to keep as it as is currently, if possible, due to these reasons.

vtjnash replied:
The benchmark is for an older commit which had a bug.

simonbyrne replied:
ok, then we should get rid of the `^ = Base.FastMath.pow_fast` if it isn't needed.

tkelman replied:
any way to add a test more visible than the perf tracking that the constant propagation is working reliably?

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19890#issuecomment-278726061) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/3eff07a82690bb404e65a9d4eadae96ed5388818/4939f20_vs_3bc9a16/report.md). cc @jrevels

musm commented:
Is the only thing left to do here is to drop the last commit, revert the exp function changes, and rerun nanosoldier?

stevengj commented:
I don't think "correctness" is the right way to frame this question.  Exponentiation is not guaranteed to be rounded to 1 ulp, so it's a question of slightly more vs. slightly less roundoff error.

The question is, is the marginal increase in accuracy worth it for the decrease in performance here?  (Exponentiation-by-squaring and similar algorithms for `x^p` have O(√log(p)) rms error growth, I believe, so the error is quite reasonable even for large `p`.)

stevengj commented:
My impression is that Fortran compilers have been inlining literal integer powers as repeated multiplications for decades, and I've never heard of anyone complaining about the accuracy. e.g. see the [gfortran code for this](gfor_fndecl_math_powi).

stevengj commented:
Ah, looking more carefully at the gfortran code, it only does the power expansion for float types for powers `>2` or `<-1` if `-funsafe-math-optimizations` is set and `-Os` is not set. See [this line](https://github.com/gcc-mirror/gcc/blob/01991e362c8d4751e97cb80fa23c2f0d1478387f/gcc/fortran/trans-expr.c#L2879-L2881).

stevengj commented:
And I guess our @fastmath already calls `powi`.

StefanKarpinski commented:
It seems like there ought to be more efficient ways to compute `x^p` precisely for floating-point `x` and fixed `p` than calling the general `pow` function.

simonbyrne commented:
Probably not for arbitrary `x`: `pow(x,y)` is typically implemented as a slightly-extended precision version of `exp(y*log(x))`.

vtjnash committed:
only use accurate powf function

The powi intrinsic optimization over calling powf is that it is inaccurate.
We don't need that.

When it is equally accurate (e.g. tiny constant powers),
LLVM will already recognize and optimize any call to a function named `powf`,
and produce the same speedup.

fix #19872

vtjnash committed:
update DomainError showerror handling to ignoring inlining

tkelman commented:
@nanosoldier `runbenchmarks(ALL, vs=":master")`

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19890#issuecomment-281791383) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/b3e401bf172f8109d6547d171ffd8de7aeaa09f2/9973b8c_vs_0cebddf/report.md). cc @jrevels

tkelman commented:
so the simple version regressed - is that the Val literal lowering change making inlining or constant prop less predictable?

vtjnash commented:
Moving the value into the type domain can potentially thwart constant folding. In this case, it's this constant-value function that got broken:
```
exp_small_thres(::Type{Float64}) = 2.0^-28
```

vtjnash committed:
make several pure-marked math functions actually pure

vtjnash committed:
make sure that the indirection through the `Val{p}` type doesn't stop inlining of `^`

vtjnash commented:
We should really stop dispatching on `Type{Val{...}}`. Working with types is much harder on inference and the runtime than `Val{...}`. Constructing the value is effectively free (it's a singleton, so it's already constructed), and it's much more likely to be properly constant folded and dispatched. Plus there are a lot of fast paths in the runtime that can be accessed by a singleton which cannot exist for types (They instead often get handed off to the very slow fallback code).

tkelman commented:
were you waiting for something before running @nanosoldier `runbenchmarks(ALL, vs=":master")` again?

JeffBezanson commented:
+100 https://github.com/JuliaLang/julia/pull/19890#issuecomment-282091331

The only real problem is that `Val{x}()` is a bit long. I believe we could define

```
@pure Val(x) = Val{x}()
```

which would allow it to be constant-folded and inferred just as well, so you could write `Val(3)`.

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/19890#issuecomment-282168390) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/729b08f059470aa27c8d55b9c2bf25e2d3a5b3d8/2dee700_vs_7968717/report.md). cc @jrevels

vtjnash commented:
> `@pure Val(x) = Val{x}()`

true. that then just has the problem that calling the `Val(x)` constructor requires the slow path :P

I'll also just reiterate that the fact there's a number of methods `foo(::Type{Val{true}})` is just completely awful. Just call it `foo_true()` if you're really hate your code reader that much.

JeffBezanson commented:
Agreed; we should not have public APIs with `Val{true}` and `Val{false}`. A rare case where a name with an underscore is less ugly.

vtjnash commented:
```
~/julia/base$ grep -RI 'Val{true}\|Val{false}' . | wc -l
      68
```

anyways, enough OT - I think we can go ahead and merge this?

JeffBezanson commented:
Yes, seems like it's time.

stevengj commented:
The constant-folding of powers is fixed by #20783, following the suggestion to use `Val{p}()`.

musm commented on base/math.jl:
```diff
@@ -20,13 +20,23 @@ import Base: log, exp, sin, cos, tan, sinh, cosh, tanh, asin,
+    @eval exponent_max(::Type{$T}) = $(Int(exponent_mask(T) >> significand_bits(T)) - exponent_bias(T))
+    # maximum float exponent without bias
+    @eval exponent_raw_max(::Type{$T}) = $(Int(exponent_mask(T) >> significand_bits(T)))
+end
```
@vtjnash  Out of curiosity, what was the reason behind moving these functions from `float.jl` to `math.jl`?  I had some code that imported `Base.significand_bits`, and now it needs `Base.Math.significand_bits` to work on master.

vtjnash replied:
They were incorrectly marked `@pure`, and only used here. You can import them from `Base.Math` on old code also, to be compatible.
